packages/grafana-runtime/src/services/ScopesContext.ts

```diff
@@ -0,0 +1,47 @@
+import { createContext, useContext } from 'react';
+import { useObservable } from 'react-use';
+import { Observable } from 'rxjs';
+
+import { Scope } from '@grafana/data';
+
+export interface ScopesContextValue {
+  state: {
+    isDrawerOpened: boolean;
+    isEnabled: boolean;
+    isLoading: boolean;
+    isReadOnly: boolean;
+    value: Scope[];
+  };
+  stateObservable: Observable<ScopesContextValue['state']>;
+  changeScopes: (scopeNames: string[]) => void;
+  enterReadOnly: () => void;
+  exitReadOnly: () => void;
+  toggleDrawer: () => void;
+  openDrawer: () => void;
+  closeDrawer: () => void;
+  enable: () => void;
+  disable: () => void;
+}
```

Comment:
just spend 10min on thinking if all this is needed and the naming etc, as changing this will be very difficult once plugins start using it

Reply:
Some thoughts I already had about it:
```
export interface ScopesContextValue {
  // I would keep it like this
  state: {
    drawerOpened: boolean;
    enabled: boolean;
    loading: boolean;
    readOnly: boolean;
    value: Scope[];
  };

  // We can adhere to state$ that is usually used in rxjs world but I haven't seen it in Grafana
  stateObservable: Observable<ScopesContextValue['state']>;

  // This one I would keep like this
  changeScopes(scopeNames: string[]): void;

  // Helpers for changeable state props
  setReadOnly(readOnly: boolean): void;
  setDrawerOpened(drawerOpened: boolean): void;
  setEnabled(enabled: boolean): void;

  // And/or we can simply pass
  setState(newState: ScopesContextValue['state']): void
}
```

WDYT about those?

Reply:
@bfmatei - i think you proposal is good. Only thing I'm not certain about, and would possibly hold on with is the `setState`. I'd rather shape the API on the user needs and extend it later on, than leaving it openended as with `setState`